netdog: Persist current IP and add subcommand `node-ip` for retrieval

zmrow opened the pull request:
**Issue number:**
Related to #1458 


**Description of changes:**
```
With this change, netdog will persist the node's current IP (read from
it's DHCP leases file) to /var/lib/netdog/current_ip.  It does this as
part of its "install" subcommand, which is run each time that wickedd
receives an update.  A new subcommand has also been added, which
retrieves the current IP from said file and returns it as JSON.  This
subcommand is intended for use as a settings generator.
```

This change is meant to support the VMWare variant #1451, specifically the `kubernetes.node-ip` setting and will receive further exercise with that PR.

**Testing done:**
* Build and run an `aws-dev` variant to ensure that the file containing the IP exists with the correct data and the subcommand works:
```
bash-5.0# ls -lahZ /var/lib/netdog/current_ip 
-rw-r--r--. 1 root root system_u:object_r:lease_t:s0 11 Apr 19 20:15 /var/lib/netdog/current_ip

bash-5.0# cat /var/lib/netdog/current_ip 
10.0.60.220

bash-5.0# netdog node-ip
"10.0.60.220"
```
* Build and run an `aws-k8s-1.19` node, started sheltie and checked files.  Node joins the cluster just fine and I can run pods on it.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

bcressey commented on sources/Cargo.lock:
```diff
@@ -2088,6 +2085,7 @@ dependencies = [
  "rand 0.8.3",
  "regex",
  "serde",
+ "serde_json",
```
I expect it'd work with just this change to `Cargo.lock` and not all the rest of it. Can you try that?

webern replied:
Was this a cargo update?

zmrow replied:
It wasn't meant to be.  I'm going to implement @bcressey 's ask above to avoid it.

bcressey commented on sources/api/netdog/src/main.rs:
```diff
@@ -299,11 +327,27 @@ fn remove(args: &Args) -> Result<()> {
-    match args.sub_command {
-        SubCommand::Install => install(&args)?,
-        SubCommand::Remove => remove(&args)?,
+    match parse_args(env::args())? {
+        (SubCommand::NodeIp, None) => node_ip()?,
+        (subcommand, Some(args)) => match subcommand {
+            SubCommand::Install => install(&args)?,
+            SubCommand::Remove => remove(&args)?,
+            SubCommand::NodeIp => {
+                usage_msg(format!("Subcommand 'node-ip' doesn't support arguments"))
+            }
+        },
+        (_, None) => usage_msg(format!("Subcommands 'install/remove' require arguments")),
```
nit: this might be clearer overall, since it doesn't mix `NodeIp` in with the handling of `Install` and `Remove`:

```suggestion
    match parse_args(env::args())? {
        (SubCommand::NodeIp, None) => node_ip()?,
        (SubCommand::NodeIp, Some(_) => usage_msg(format!("Subcommand 'node-ip' doesn't support arguments")).
        (subcommand, Some(args)) => match subcommand {
            SubCommand::Install => install(&args)?,
            SubCommand::Remove => remove(&args)?,
        },
        (_, None) => usage_msg(format!("Subcommands 'install/remove' require arguments")),
```

bcressey replied:
```
(_, None) => usage_msg(format!("Subcommands 'install/remove' require arguments"))
```

For this it might be better to use the `subcommand` parameter so you can add it to the format string.

webern replied:
I suggest:

```rust
enum Subcommand {
    Install(Args),
    Remove(Args),
    NodeIp,
}
```

And

```rust
fn parse_args(args: env::Args) -> Result<SubCommand> {
   // ...
}
```

webern replied:
Actually since, since this would require us to stop using serde_plain to 'deserialize' the subcommand string, it's probably not worth it.

zmrow replied:
@bcressey - The compiler wasn't smart enough to do your suggestion, it still wanted an exhaustive match of the enum variants inside the second `match`.  I opted to split out the options a bit further and just avoid a nested match altogether.

bcressey reviewed (commented):
Can you also verify that `/etc/resolv.conf` and `/proc/sys/kernel/hostname` are still set correctly?

webern reviewed (commented):
Is persisting the IP to a file an optimization (i.e. for faster queries)? Or is the persistence needed across reboots?

zmrow commented:
@webern Persisting the IP to file means that we can use it as the static IP for the setting `kubernetes-node-ip`, specifically for VMWare (since we don't have the luxury of IMDS).

zmrow committed:
netdog: Persist current IP and add subcommand `node-ip` for retrieval

With this change, netdog will persist the node's current IP (read from
it's DHCP leases file) to /var/lib/netdog/current_ip.  It does this as
part of its "install" subcommand, which is run each time that wickedd
receives an update.  A new subcommand has also been added, which
retrieves the current IP from said file and returns it as JSON.  This
subcommand is intended for use as a settings generator.

zmrow commented:
^ Addressed @bcressey 's comments and verified that `/etc/resolv.conf` and `/proc/sys/kernel/hostname` are both still set correctly.
